xdp/pp_poll.c

```diff
@@ -0,0 +1,353 @@
+#include "common.h"
+#include "src/net.h"
+#include "src/persistence.h"
+#include "src/xdp-loading.h"
+
+#include <stdbool.h>
+#include <stdlib.h>
+
+void usage (char *prog)
+{
+    // <prog> <ifname> <action>
+    fprintf (stderr, "Usage: %s <ifname> <action> [extra arguments]\n", prog);
+    fprintf (stderr, "Actions:\n");
+    fprintf (stderr, "\t- start: start the pingpong experiment\n");
+    fprintf (stderr, "\t         on the server: %s <ifname> start <# of packets>\n", prog);
+    fprintf (stderr, "\t         on the client: %s <ifname> start <# of packets> <packets interval (ns)> <server IP>\n", prog);
+    fprintf (stderr, "\t- remove: remove XDP program\n");
+}
+
+// Information about the XDP program
+static const char *filename = "pingpong.o";
+static const char *prog_name = "xdp_main";
+static const char *pinpath = "/sys/fs/bpf/xdp_pingpong";
+static const char *mapname = "last_payload";
+
+static const char *outfile = "pingpong.dat";
+
+static persistence_agent_t *persistence;
+
+// global variable to store the loaded xdp object
+static struct bpf_object *loaded_xdp_obj;
+
+/**
+ * Poll the map until a new payload is found.
+ * This function busy-waits until a new payload is found.
+ * When a new payload is found, it is copied to the given destination payload and the map entry is cleared.
+ *
+ * @param map_ptr the pointer to the map
+ * @param dest_payload the pointer to the payload to be filled
+ * @param last_index the index of the last payload retrieved
+ * @param busy_poll_next if true, the function busy-waits the next index, otherwise round-robin is used
```

Comment:
What do you mean by "round-robin" here? Does it not just go loop back to the start since we are looking at the last element?

Reply:
While writing the reason behind `busy_poll_next` I realized it is not even needed anymore. 
In the current way, there is no need for `busy_poll_next`: xdp keeps track of last insertion place in the map, and puts the new packet after that one; userspace app keeps track of the latest read position, and resumes reading from the next position. They basically follow each other, so if `map[curr_idx]` does not contain a packet, then there must be no other packet in the map.
I am struggling to recall the reason for `busy_poll_next` to be there; I am sure it was required because the last `map_size` packets received by the machine would not be polled, so the function needed to be told to go through the whole array looking for new packet, not just the next position.
I will remove the flag and related code in the next commits.

Reply:
Sounds good. I'll let you choose whether you want to do that in this PR or in a future one